documentation/course/e2/README.md

```diff
@@ -2047,7 +2047,36 @@ TODO @jakubzehner: Dodać diagram pakietów, opis architektury i endpointy.
 
 ## Płatność
 
-TODO @piterek130: Dodać diagram pakietów, opis architektury i endpointy.
+TODO @piterek130: Dodać diagram pakietów, opis architektury.
+
+### API
+
+#### API publiczne
+
+| **Rola**                 | **Metoda** | **Endpoint**                   | **Wymagania**                | **Opis**                                    |
+| ------------------------ | ---------- | ------------------------------ | ---------------------------- | ------------------------------------------- |
+| `passenger`              | `GET`      | `/ext/v1/cards`                | `PAY/19`                     | Pobranie listy kart płatniczych.            |
+| `passenger`              | `POST`     | `/ext/v1/cards`                | `PAY/02`                     | Dodanie nowej karty płatniczej.             |
+| `passenger`              | `PUT`      | `/ext/v1/cards/:id`            | `PAY/20`                     | Zaktualizowanie danych karty płatniczej.    |
+| `passenger`              | `DELETE`   | `/ext/v1/cards/:id`            | `PAY/03`                     | Usunięcie karty płatniczej.                 |
+| `passenger`              | `POST`     | `/ext/v1/wallet/add-funds`     | `PAY/07`                     | Doładowanie portfela.                       |
+| `passenger`              | `GET`      | `/ext/v1/wallet`               | `PAY/08`                     | Pobranie stanu portfela.                    |
+| `passenger`              | `GET`      | `/ext/v1/wallet/history`¹      | `PAY/09`                     | Pobranie historii doładowań portfela.       |
+| `passenger`              | `GET`      | `/ext/v1/transations`¹         | `PAY/10`                     | Pobranie historii transakcji.               |
+| `passenger`, `inspector` | `GET`      | `/ext/v1/fines`¹               | `PAY/17`                     | Pobranie listy mandatów.                    |
+| `passenger`              | `GET`      | `/ext/v1/fines/:id`            | `PAY/18`                     | Pobranie informacji o mandacie.             |
+| `passenger`              | `POST`     | `/ext/v1/payment`              | `PAY/14`, `PAY/15`, `PAY/16` | Realizacja płatności.                       |
```

Comment:
**suggestion:** Okay, there's this weird trick, which you should probably know about, since it will simplify a few things:

You can merge `/ext/v1/payment` and `/ext/v1/wallet/add-funds` into one endpoint if you move some logic into the frontend. More specifically, currently (AKA **before**):
- to charge wallet: `/wallet/add-funds` (supports all payment methods)
- to pay for a ticket: `/payment` (supports all payment methods)
- to pay a fine: ??? (or are you using `/payment` for both?)

Note how all of these have duplicate logic to handle all payment methods.

You can instead (**after**):
- to charge wallet: `/wallet/add-funds` (supports all payment methods)
- to pay for a ticket: `/wallet/add-funds` -> Kuba's ticket endpoint (which calls your `/charge`)
- to pay for a fine: `/wallet/add-funds` -> `/fines/:id/pay` (which supports only wallet payments)

Then you only have to support different payment methods in `/wallet/add-funds`, and the other endpoints just charge the wallet, which is a simple subtraction of a number in our database and fail if the amount in wallet is insufficient. You don't have to do duplicate error handling logic and integration with the payment gateway, etc.

To do this on frontend you have the exact same components as before, you can select payment methods both when charging the wallet, paying a fine and buying a ticket, except when clicking "buy ticket" with "credit card" as a payment method, you call `/wallet/add-funds` regardless.

Not sure if I described it correctly, also cc @jakubzehner since he would be the person who's life could get harder with this change and also the person who knows what I'm talking about since we implemented it together some time ago.

Reply:
That's right, it simplifies a lot of things, just make the frontend send more calls.